src/BatteryStateOfChargeService.cpp

```diff
@@ -43,21 +48,55 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    if (totalAH_>BATTERY_AMP_HOUR_CAPACITY)
+        return 0;
+    else if (totalAH_<0)
+        return BATTERY_AMP_HOUR_CAPACITY;
+    else
+        return BATTERY_AMP_HOUR_CAPACITY - totalAH_;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    return isCharging_;
 }
 
+
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    QTime null_time(0,0,0,0);
+    if (return_time.isNull())
+        return null_time;
+    else
+        return return_time;
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
-    Q_UNUSED(batteryData);
-    // Update your variables here.
+    if (previousTime_.isNull())
+    {
+        totalAH_=BATTERY_AMP_HOUR_CAPACITY*initialStateOfChargePercent_/100;
+        previousTime_=batteryData.time;
+    }
+    zero_time.setHMS(0,0,0,0);
+    deltaTime_=previousTime_.msecsTo(batteryData.time)/HOUR_TO_SEC;
+    deltaCurrent_=(batteryData.current+previousCurrent_)/2;
+    amph_=deltaTime_*deltaCurrent_;
+    totalAH_+=amph_;
```

Comment:
The `amph_` is only used once, it would be fine to remove to make for shorter code. I would recommend just changing line 84 and 85 to:
`totalAH_ = deltaTime_ * deltaCurrent_;`. 
If you think the intermediate step is necessary to show what you are calculating, then `amph_` would only need to be a local variable and not a private member. Also, not abbreviating the name would better, so `amph_ --> ampHours_`